rust/cargo_manifest_to_build.py

```diff
@@ -0,0 +1,561 @@
+    flags="".join(["\n        \"%s\"," % f for f in context["flags"]]),
+    other=context["other"],
+    version=context["version"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+)
+
+def rust_binary(target, context):
+    return """rust_binary(
+    name = "{name}_bin",
+    crate_root = "{path}",
+    srcs = glob(["**/*.rs"]),
+    deps = [":{name}", {deps}],
+    rustc_flags = [
+        "--cap-lints allow",{flags}
+    ],{other}
+    version = "{version}",
+    crate_features = [{features}],
+)""".format(
+    name=target["name"].replace("-", "_"),
+    path=remove_path_prefix(target["src_path"], context["workspace_root"]),
+    deps=context["deps"],
+    flags="".join(["\n        \"%s\"," % f for f in context["flags"]]),
+    other=context["other"],
+    version=context["version"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+)
+
+def custom_build_script(target, context):
+    return """rust_binary(
+    name = "{name}_build_script",
+    crate_root = "{path}",
+    srcs = glob(["**/*.rs"]),
+    deps = [{build_deps}],
+    rustc_flags = ["--cap-lints allow"],
+    version = "{version}",
+    crate_features = [{features}],
+    visibility = ["//visibility:private"],
+)
+
+cargo_build_script_run(
+    name = "{name}_build_script_executor",
+    srcs = glob(["*", "**/*.rs"]),
+    script = ":{name}_build_script",
+    features = [{features}],
+)
+""".format(
+    name=context["name"].replace("-", "_"),
+    path=remove_path_prefix(target["src_path"] or "build.rs", context["workspace_root"]),
+    build_deps=context["build_deps"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+    version=context["version"],
+)
+
+def resolve_deps(deps, ctxt):
+    return ",".join([
+        "\"%s\"" % ctxt["dep_format"].format(
+            name=d["name"].replace("-", "_"),
+            version=ctxt["resolved_deps"][d["name"]].replace(".", "_")
+        )
+        for d in deps
+        if d["name"] in ctxt["resolved_deps"]
+        ])
+
+def extend_context(ctxt, json):
+    out_dir_tar = ""
+    if any([t["kind"] == ["custom-build"]] for t in json["targets"]):
+        out_dir_tar = "\n    out_dir_tar = \":%s_build_script_executor\"," % json["name"].replace("-", "_")
+    data = ""
+    if ctxt["data"]:
+        data = "\n    data = r\"\"\"%s\"\"\"," % ctxt["data"]
+    for d in json["dependencies"]:
+        if d["name"] not in ctxt["resolved_deps"]:
+            sys.stderr.write("WARNING: Cannot resolve dependency on crate '%s'\n" % d["name"])
+    return {
+        "name": json["name"],
+        "workspace_root": ctxt["workspace_root"],
+        "flags": ctxt["flags"],
+        "version": json["version"],
+        "other": "%s%s" % (out_dir_tar, data),
+        "deps": resolve_deps([d for d in json["dependencies"] if not d["kind"]], ctxt),
+        "build_deps": resolve_deps([d for d in json["dependencies"] if d["kind"] == "build"], ctxt),
+        "features": json["features"].keys(),
```

Comment:
Experimentally, I found that crate["features"] yielded from `cargo metadata` as described below is never going to have the required features, because they cannot be identified in isolation. I think they need to be supplied by the caller (of the repository_rule) on a per-target basis with the wider crate context in mind

Reply:
This behavior was copied from cargo raze. We can definitely provide a way to override it from the attributes but it would diverge from the way cargo raze works.

Reply:
There is a difference between what `cargo raze` does and what you're doing here. Cargo raze runs the metadata analysis once for the whole tree with all of the workspace context present. If I understand what you're doing here completely, you're running a similar process once per crate brought in.